Fix ambiguity of a type deduction in TEST_EQ macro for `enum class` args.

vglavnyy opened the pull request:
Before this PR this code  (in the `test.cpp`) was impossible:
```cpp
enum class TestEnum : int8_t {  A, B };
TEST_EQ(MyGame::Example::TestEnum::A, Foo());
```

Due to ambiguity of type deduction, this code caused a compilation error like this:
>error C2679: binary '<<': no operator found which takes a right-hand operand of type 'T' (or there is no acceptable conversion) with [ T=MyGame::Example::TestEnum  ]
.... (listing of all possible specializations)

This is a well-known problem (C++ DR1601):
- http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#1601
- https://stackoverflow.com/questions/34336024/ambiguous-overload-when-writing-an-enum-with-an-enum-base-but-only-with-clang

However, all modern compilers break the rule C++17 16.4.2 introduced by DR 1601:
>A conversion that promotes an enumeration whose underlying type is fixed to its underlying type is better than one that promotes to the promoted underlying type, if the two are different.

This PR adds a workaround and allows the use of simple checking of `enum class` values in tests:
```cpp
TEST_EQ(MyGame::Example::TestEnum::A, MyGame::Example::TestEnum::B);
```

aardappel commented on tests/test_assert.h:
```diff
@@ -54,14 +54,55 @@ void TestFail(const char *expval, const char *val, const char *exp,
+  #if defined(FLATBUFFERS_TEMPLATES_ALIASES) && !defined(FLATBUFFERS_CPP98_STL)
+  using type = typename std::underlying_type<T>::type;
+  #else
+  typedef int64_t type;
```
I think `int32_t` for enums was more common on older compilers, but it probably doesn't matter.

vglavnyy replied:
MSVC2010 doesn't support non-fixed enums bigger than `int`, but gcc and clang do.
But you are right, it does not matter for the tests.

aardappel commented:
Actually, CI errors appear related.

vglavnyy commented:
CI errors are connected with #5611.
```cpp
TEST_EQ(root_v1_viewed_from_v2->j(), NULL);
TEST_EQ(root_v1_viewed_from_v2->e()->c(), NULL);
TEST_EQ(root_v1_viewed_from_v2->c_as_TableC(), NULL);
```
NULL is int. I already wrote this in #5611.
I will re-run CI after fixes.

aardappel commented:
Still an Android error: https://travis-ci.org/google/flatbuffers/jobs/615311373#L2409

Looks like the STLPort emulation we have is missing `operator==` for `unique_ptr` and `nullptr` or something :(

vglavnyy commented:
>Looks like the STLPort emulation we have is missing operator== for unique_ptr and nullptr or something :(

@aardappel, fixed, all tests passed.
I'm sorry. I squashed the fix with first commit before review.
Can you review this PR again?

Changes from the initial commit:
* ` tests/test_assert.h`
  - added missed `()` int TEST_NOTNULL(),
  - added explicit cast to `bool` in TEST_ASSERT().
* `stl_emulation.h`
  - added operators for comparison with `nullptr`.

aardappel commented:
Nice, LGTM!
